<div class="alert block alert-info alert">

# <center> Scientific Programming in Python
## <center>Karl N. Kirschner<br>Bonn-Rhein-Sieg University of Applied Sciences<br>Sankt Augustin, Germany

# <center> Rounding Numbers
#### <center> (Reporting the correct number of significant figures.)</center>

#### Sources
1. https://docs.python.org/3/library/functions.html#round
2. https://docs.python.org/3/library/decimal.html#decimal.Decimal
3. https://realpython.com/python-rounding/#the-decimal-class
4. https://docs.python.org/3/library/math.html
5. https://docs.python.org/3/library/decimal.html
6. https://numpy.org/doc/stable/reference/generated/numpy.around.html
7. https://numpy.org/doc/stable/reference/generated/numpy.ceil.html
8. https://numpy.org/doc/stable/reference/generated/numpy.floor.html
9. https://docs.python.org/3.1/library/string.html#grammar-token-precision
<hr style="border:2px solid gray"></hr>

In [1]:
## For extra information given within the lectures

from IPython.display import HTML


def set_code_background(color: str):
    ''' Set the background color for code cells.

        Source: psychemedia via https://stackoverflow.com/questions/49429585/
                how-to-change-the-background-color-of-a-single-cell-in-a-jupyter-notebook-jupy

        To match Jupyter's dev class colors:
            "alert alert-block alert-warning" = #fcf8e3

        Args:
            color: HTML color, rgba, hex
    '''

    script = ("var cell = this.closest('.code_cell');"
              "var editor = cell.querySelector('.input_area');"
              f"editor.style.background='{color}';"
              "this.parentNode.removeChild(this)")
    display(HTML(f'<img src onerror="{script}">'))


set_code_background(color='#fcf8e3')

#### <center> Don't Scroll Down Yet...</center>

**Question**: How would you round -2.45 to the first decimal place?

<!-- (Everyone type in their number in the "Chat box.") -->
<br><br><b>
<br><br><b>
<br><br><b>

<br><br><b>
<br><br><b>
<br><br><b>

<br><br><b>
<br><br><b>
<br><br><b>

<br><br><b>
<br><br><b>
<br><br><b>

<br><br><b>
<br><br><b>
<br><br><b>

<br><br><b>
<br><br><b>
<br><br><b>

<br><br><b>
<br><br><b>
<br><br><b>

<hr style="border:2px solid gray"></hr>

## Conceptual ideas

- Critical thinking is needed when rounding numbers whose last digit is a <font color='dodgerblue'>**5**</font> since it lies in the middle of the basic number series <br> (i.e. **<font color='Brown'>0, 1, 2, 3, 4</font> -- <font color='dodgerblue'>5</font> -- <font color='Brown'>6, 7, 8, 9, 10</font>**). <br><br> So...should it be associated with the smaller or larger numbers?


- The act of rounding can introduce a bias into your numbers (e.g. when working with a large data set).

Consider 11 numbers that range from 0.5 -- 10.5 in 1.0 increments
- **round them** to the one's place (**manually** - by hand), and
- then take their **mean**.
    
We will use that statistics library since it allows us to compute the mean.

In [2]:
import statistics

In [3]:
input_numbers = [number+0.5 for number in range(0, 11, 1)]

print(f'Unrounded Numbers: {input_numbers}')
print(f'Average: {statistics.mean(input_numbers)}')

Unrounded Numbers: [0.5, 1.5, 2.5, 3.5, 4.5, 5.5, 6.5, 7.5, 8.5, 9.5, 10.5]
Average: 5.5


Rounding the number up (manually):

In [4]:
round_up = [number for number in range(1, 12, 1)]

print(f'Rounded Up Numbers: {round_up}')
print(f'Mean: {statistics.mean(round_up)}')

Rounded Up Numbers: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
Mean: 6


Rounding the number down (manually):

In [5]:
round_down = [number for number in range(0, 11, 1)]

print(f'Rounded Down Numbers: {round_down}')
print(f'Mean: {statistics.mean(round_down)}')

Rounded Down Numbers: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Mean: 5


Thus, we observe that:
1. Rouding Up leads to a **mean of 6**, while
2. Rounding Down leads to a **mean of 5**.

<font color='dodgerblue'>**Therefore, the approach that we use to round numbers may result in a bias.**</font>

<hr style="border:2px solid gray"></hr>

## Conceptual approaches for rounding numbers

### 1. Bias methods

<font color='dodgerblue'>**Round Half Up**</font> (i.e. bias towards **more positive**)
1.  7.6 $\rightarrow$ 8
2. **7.5 $\rightarrow$ 8**
3. 7.4 $\rightarrow$ 7
<br><br>
4. -7.6 $\rightarrow$ -8
5. **-7.5 $\rightarrow$ -7**
6. -7.4 $\rightarrow$ -7

<font color='dodgerblue'>**Round Half Down**</font> (i.e. bias towards **more negative**)
1.  7.5 $\rightarrow$  7
2. -7.5 $\rightarrow$ -8

<font color='dodgerblue'>**Round Half To Zero**</font> (i.e. bias towards **zero**)
1.  7.5 $\rightarrow$  7
2. -7.5 $\rightarrow$ -7

<font color='dodgerblue'>**Round Half Away from Zero**</font> (i.e. bias towards **$\pm \infty$**)
1. 7.5 $\rightarrow$  8
2. -7.5 $\rightarrow$ -8

Additional information:

https://en.wikipedia.org/wiki/Rounding

### 2. Unbias methods (i.e. free from positive/negative bias and bias toward/away from zero)

<font color='dodgerblue'>**Round Half To Even**</font> (i.e. zero is considered evenly distributed)
1.  7.5 $\rightarrow$  8
2.  8.5 $\rightarrow$  8
<br><br>
3. -7.5 $\rightarrow$ -8
4. -8.5 $\rightarrow$ -8


<font color='dodgerblue'>**Round Half To Odd**</font> (i.e. zero is considered evenly distributed)
1.  7.5 $\rightarrow$  7
2.  8.5 $\rightarrow$  9
<br><br>
3. -7.5 $\rightarrow$ -7
4. -8.5 $\rightarrow$ -9

#### Visual clarification of bias and unbias methods

![rounding](00_images/rounding_bias.png)

<hr style="border:2px solid gray"></hr>

## Different approaches within Python

### 1. Python's built-in function `round`

- https://docs.python.org/3/library/functions.html#round

Using the numbers that are exemplified in the above figure:

In [6]:
example_numbers = [-4.5, -3.5, -0.5, 0.5, 3.5, 4.5]

for number in example_numbers:
    number_rounded = round(number)

    print(f'The number {number} rounds to {number_rounded}.')

The number -4.5 rounds to -4.
The number -3.5 rounds to -4.
The number -0.5 rounds to 0.
The number 0.5 rounds to 0.
The number 3.5 rounds to 4.
The number 4.5 rounds to 4.


**Thus**, Python3's built-in **`round`** function uses <font color='dodgerblue'>**Round Half To Even**</font> to the ones place (by default).
- From the doc's page: "... rounding is done toward the even choice ..."
<br><br>


What happens if we push it slightly over 4.5?

In [7]:
print(round(4.51))

5


So, the behavior is has it should be:  since the value is a **little greater** than four-and-a-half, it should **round up**.

**Notice** how `round` simply returns the integer value of the float above.

Alternatively, one can specify the number of **decimal places** using **`round(number[, ndigits])`**.
- However, this does odd things to the number of digits that are returned (i.e. significant figures).

Let's round a list of floats to the second decimal place (i.e. to have 1 sigfig):

In [8]:
numbers_demo_expected = [-0.055, 0.055]

for number in numbers_demo_expected:
    number_rounded = round(number, 2)
    
    print(f'When rounded to 2 decimals, the number {number} rounds to: {number_rounded}')

When rounded to 2 decimals, the number -0.055 rounds to: -0.06
When rounded to 2 decimals, the number 0.055 rounds to: 0.06


The results are as expected.

In [9]:
numbers_demo_questionable = [10, 0.055E-9]

for number in numbers_demo_questionable:
    number_rounded = round(number, 2)
    
    print(f'When rounded to 2 decimals, the number {number} rounds to: {number_rounded}')

When rounded to 2 decimals, the number 10 rounds to: 10
When rounded to 2 decimals, the number 5.5e-11 rounds to: 0.0


Notice two things:

1. **10**, which as no decimals initially -- and thus a uncertain number of significant figures -- stays as 10 (which is good)

2. The function `round` returned **0.0** for **0.055E-9**
    - the scientifically correct answer is <font color='dodgerblue'>**0.06E-9**</font> (or 0.6E-10 or 6.E-11, depending on the numbering format convention)
    - which means we <font color='dodgerblue'>**lost**</font> some data information

**Furthermore**, notice the following weird result when we round 2.675 to two decimal places.

In [10]:
print( round(2.675, 2) )

2.67


The `round(2.675, 2)` should result in **2.68** (i.e. rounding up). Instead we obtained **2.67**!
<br><br>

Two more examples of weirdness that helps us understand what is happening:

In [11]:
0.1 + 0.1 + 0.1

0.30000000000000004

In the following, we are adding numbers that should result in zero (i.e., 0.0):

In [12]:
(0.1 + 0.1 + 0.1) - 0.3

5.551115123125783e-17

Why?

Reason: <font color='dodgerblue'>**floating-point representation error**</font> (https://docs.python.org/3/tutorial/floatingpoint.html)

This is not a Python issue, but has to do with how computers store floating-point numbers in memory.

**Take-home message:**

- From a perspective of being exact and correct, this is something that one needs to be aware of -- particularly when significant figures are important -- when developing code for scientific usage.

<hr style="border:2px solid gray"></hr>

### 2. Python's math library (with examples)

- https://docs.python.org/3/library/math.html

- `math.ceil` (ceiling): **Round Up**
- `math.floor`: **Round Down**

Both functions return only an **integer** value.

In [13]:
import math

example_numbers = [-4.5, -3.5, -0.5, 0.5, 3.5, 4.5]

In [14]:
for number in example_numbers:
    number_rounded = math.ceil(number)

    print(f'The number {number} rounds to {number_rounded}.')

The number -4.5 rounds to -4.
The number -3.5 rounds to -3.
The number -0.5 rounds to 0.
The number 0.5 rounds to 1.
The number 3.5 rounds to 4.
The number 4.5 rounds to 5.


`math.ceil` rounds all numbers to more <font color='dodgerblue'>positive</font> values.

In [15]:
for number in example_numbers:
    number_rounded = math.floor(number)

    print(f'The number {number} rounds to {number_rounded}.')

The number -4.5 rounds to -5.
The number -3.5 rounds to -4.
The number -0.5 rounds to -1.
The number 0.5 rounds to 0.
The number 3.5 rounds to 3.
The number 4.5 rounds to 4.


`math.floor` rounds all numbers to more <font color='dodgerblue'>negative</font> values.

<hr style="border:2px solid gray"></hr>

### 3. The decimal library
- https://docs.python.org/3/library/decimal.html
    - "The decimal module provides support for fast correctly-**rounded decimal floating point arithmetic**."
    - Can be passed a number that is either a float (e.g. `3.14`, or a string (e.g. `'3.14'`)
        - The <font color='dodgerblue'>string approach</font> allows us to correct for the <font color='dodgerblue'>floating-point representation error</font>
    - Rounding is done with `decimal.Decimal.quantize`
    - Returns a `decimal.Decimal` object

Rounding options:
1. ROUND_CEILING $\rightarrow$ Round towards positive infinity
2. ROUND_FLOOR $\rightarrow$ Round towards negative infinity 
3. ROUND_DOWN $\rightarrow$ Round towards zero
4. ROUND_UP $\rightarrow$ Rounding away from zero
5. ROUND_HALF_UP $\rightarrow$ Rounding half away from zero
6. ROUND_HALF_DOWN $\rightarrow$ Rounding half towards zero
7. ROUND_HALF_EVEN (<font color='dodgerblue'>default</font>) $\rightarrow$ Rounding half to even
8. ROUND_05UP $\rightarrow$ Rounding towards zero, unless the last digit is 0 or 5 then round away from zero

Notice that the <font color='dodgerblue'>default</font> rounding method is **ROUND_HALF_EVEN**.

In [16]:
import decimal
from decimal import Decimal

Let's first consider how `decimal.Decimal` handles
- **floats** and
- **strings**

In [17]:
Decimal(3.14)

Decimal('3.140000000000000124344978758017532527446746826171875')

In [18]:
Decimal('3.14')

Decimal('3.14')

In [19]:
type(Decimal('3.14'))

decimal.Decimal

Now let's look at what happens if we do some mathematics (i.e. add the number to itself).

We should get:

3.14 + 3.14 = 6.28

In [20]:
Decimal(3.14) + Decimal(3.14)

Decimal('6.280000000000000248689957516')

In [21]:
Decimal('3.14') + Decimal('3.14')

Decimal('6.28')

Therefore, using the **string input** approach with **`decimal.Decimal`** <font color='dodgerblue'>**corrects**</font> the <font color='dodgerblue'>**floating-point representation error**</font>.

#### Rounding with decimals library

- rounging done using ```quantized```:
    - ```Decimal(value="0", context=None).quantize(exp, rounding=None, context=None)```


- Decimal().quantize('0') $\rightarrow$ ones place (e.g. 7)
- Decimal().quantize('0.0') $\rightarrow$ ones and first decimal place (e.g. 7.1)
- Decimal().quantize('0.00') $\rightarrow$ ones and first decimal place (e.g. 7.11)

In [22]:
# help(decimal.Decimal.quantize)

In our first example, let do the following:
1. round to the one's place (i.e. `quantize(Decimal('0.')`), and
2. pass the number as a **string** (via `f'{number}'`).

The default setting is **ROUND_HALF_EVEN**.

In [23]:
example_numbers = [-4.5, -3.5, -0.5, 0.5, 3.5, 4.5]

for number in example_numbers:
    number_rounded = Decimal(f'{number}').quantize(Decimal('0.'))

    print(f'The number {number} rounds to: {number_rounded}')

The number -4.5 rounds to: -4
The number -3.5 rounds to: -4
The number -0.5 rounds to: -0
The number 0.5 rounds to: 0
The number 3.5 rounds to: 4
The number 4.5 rounds to: 4


Let's be more exact by specifying the rounding directly, and not "hide" what is being done.

Notice that the **code below** allows one a <font color='dodgerblue'>quicker understanding</font> of what is happening in comparison to the above code (assuming that you see this code for the first time). You are saving the effort and time of future code readers $\rightarrow$ more **transparent**.

In [24]:
example_numbers = [-4.5, -3.5, -0.5, 0.5, 3.5, 4.5]

for number in example_numbers:
    number_rounded = Decimal(value=f'{number}').quantize(Decimal(value='0.'), rounding='ROUND_HALF_EVEN')

    print(f'The number {number} rounds to: {number_rounded}')

The number -4.5 rounds to: -4
The number -3.5 rounds to: -4
The number -0.5 rounds to: -0
The number 0.5 rounds to: 0
The number 3.5 rounds to: 4
The number 4.5 rounds to: 4


Now **change** the **number of decimal places** to return.

1 decimal place (i.e. `Decimal(value='0.0'`):

In [25]:
example_demo = [-4.55, -4.65, 4.55, 4.65]

for number in example_demo:
    number_rounded = Decimal(value=f'{number}').quantize(Decimal(value='0.0'), rounding='ROUND_HALF_EVEN')

    print(f'The number {number} rounds to: {number_rounded}')

The number -4.55 rounds to: -4.6
The number -4.65 rounds to: -4.6
The number 4.55 rounds to: 4.6
The number 4.65 rounds to: 4.6


Now let's change our rounding method to **rounding half up** (i.e. <font color='dodgerblue'>rounding half away from zero</font>).

In [26]:
example_numbers = [-4.5, -3.5, -0.5, 0.5, 3.5, 4.5]

for number in example_numbers:
    number_rounded = Decimal(value=f'{number}').quantize(Decimal(value='0.'), rounding='ROUND_HALF_UP')

    print(f'The number {number} rounds to: {number_rounded}')

The number -4.5 rounds to: -5
The number -3.5 rounds to: -4
The number -0.5 rounds to: -1
The number 0.5 rounds to: 1
The number 3.5 rounds to: 4
The number 4.5 rounds to: 5


**Side note**: earlier we saw that `round(2.675, 2)` resulted in **2.67**, not the correct value of **2.68**. Let's see what `Decimal.quantize` does:

In [27]:
round(2.675, 2)

2.67

In [28]:
Decimal(value="2.675").quantize(Decimal(value="0.00"), rounding='ROUND_HALF_UP')

Decimal('2.68')

Therefore, the exact answer is now obtained (i.e. without the influence of floating point representation error).

<!-- #### Precision

Let now change the default **precision**. Precision here seem to be how many **digits** (i.e. not decimal places) are **available** for a number (default value is 28). **Note** that a number doesn't need to use as 28 digits, but they are there if needed.

Decimal('11') / Decimal('7')

Let's verify the precision by computing how characters the number has, which includes the decimal point:

len(str(Decimal('11') / Decimal('7')))

decimal.getcontext().prec=3
Decimal('11') / Decimal('7')

**Be careful**: This doesn't always alter the default settings of decimal. On my installed Jupyter-notebook (bug?), the default decimals revert back to 28. However, on Colaboratory the above command does alter the default settings.

Decimal('11') / Decimal('7')

decimal.getcontext()

-->

<hr style="border:2px solid gray"></hr>

### 4. Rounding (maybe) by formating print statements containing floats

https://docs.python.org/3.1/library/string.html#grammar-token-precision


- `g`: General format
    - fixed point or exponential, depending on the number's magnitude

In [33]:
print(f'{3.4500045:g}, {3.5500055:g}, {300000000005:g}')

3.45, 3.55001, 3e+11


- **`f`**: Fixed point
    - unclear rounding

In [34]:
demo_numbers = [3.44, 3.56, -3.44, -3.56, 300000000005.44]

In [35]:
for number in demo_numbers:
    print(f'{number:0.1f}')

3.4
3.6
-3.4
-3.6
300000000005.4


What happens if we change `1f` to `3f`?

In [36]:
for number in demo_numbers:
    print(f'{number:0.3f}')

3.440
3.560
-3.440
-3.560
300000000005.440


Considering significant figures, we are **introducing new information** when it was originally **not present**.

<br><br>

Let's see if we can figure out what rules fixed point formatting is doing:

In [37]:
example_numbers = [-4.5, -3.5, -0.5, 0.5, 3.5, 4.5]

for number in example_numbers:
    print(f'The number {number} rounds to {number:0.0f}.')

The number -4.5 rounds to -4.
The number -3.5 rounds to -4.
The number -0.5 rounds to -0.
The number 0.5 rounds to 0.
The number 3.5 rounds to 4.
The number 4.5 rounds to 4.


Seems to be **round to even**.

However, consider the following:

In [38]:
example_demo = [-4.55, -4.65, -0.05, 0.05, 4.55, 4.65]

for number in example_demo:
    print(f'The number {number} rounds to: {number:0.1f}')

The number -4.55 rounds to: -4.5
The number -4.65 rounds to: -4.7
The number -0.05 rounds to: -0.1
The number 0.05 rounds to: 0.1
The number 4.55 rounds to: 4.5
The number 4.65 rounds to: 4.7


Now the behavior of formatting stings is **round to odd**. Weird.

- **`e`**: Exponent notation
    - rounding to odd number

In [39]:
print(f'{3.45:0.1e}, {3.55:0.1e}, {355000000000:0.1e}')

3.5e+00, 3.5e+00, 3.6e+11


However, I'm not sure why 355000000000 $\rightarrow$ 3.6e+11 (i.e., **round to even**) and not 3.5e+11 (i.e., **round to odd**)

In [40]:
example_demo = [-4.55, -4.65, -0.05, 0.05, 4.55, 4.65]

for number in example_demo:
    print(f'The number {number} rounds to: {number:0.1e}')

The number -4.55 rounds to: -4.5e+00
The number -4.65 rounds to: -4.7e+00
The number -0.05 rounds to: -5.0e-02
The number 0.05 rounds to: 5.0e-02
The number 4.55 rounds to: 4.5e+00
The number 4.65 rounds to: 4.7e+00


Notice how **0.05** $\rightarrow$ **5.0e-02** (i.e. a **gain of 1 significant figure**).

Even worse number reporting is in the examples below:
1. 1.55$\rightarrow$ okay, doesn't change

2. 15 $\rightarrow$ 15.00 (i.e. a gain of significant figures)

3. 1.555e-9 $\rightarrow$ 0.00 (ie. a loss of significant figures)

In [41]:
number_list = [1.55, 15, 1.555E-9]

for number in number_list:
    print(f'When rounded to 2 decimals, the number {number} rounds to: {number:0.2f}')

When rounded to 2 decimals, the number 1.55 rounds to: 1.55
When rounded to 2 decimals, the number 15 rounds to: 15.00
When rounded to 2 decimals, the number 1.555e-09 rounds to: 0.00


Let's revisit what happens with **`g`**:

In [42]:
for number in number_list:
    print(f'When using general formating, the number {number} rounds to: {number:0.2g}')

When using general formating, the number 1.55 rounds to: 1.6
When using general formating, the number 15 rounds to: 15
When using general formating, the number 1.555e-09 rounds to: 1.6e-09


1. 1.55$\rightarrow$ okay, rounds up to first decimal place

2. 15 $\rightarrow$ 15 okay, leaves it the same (i.e., no no information)

3. 1.555e-9 $\rightarrow$ 1.6e-09 okay, rounds up to first decimal place


So at least that is a little better - but one still needs to be careful...

<div class="alert alert-block alert-warning">
<hr style="border:1.5px dashed gray"></hr>

### The Numpy library

(For consideration in the future when we do cover Numpy)

- Rounding is done using
    - `numpy.around` https://numpy.org/doc/stable/reference/generated/numpy.around.html - **round to even method**
    - `numpy.ceil` (https://numpy.org/doc/stable/reference/generated/numpy.ceil.html) - rounding up (i.e., more positive)
    - `numpy.floor` (https://numpy.org/doc/stable/reference/generated/numpy.floor.html) - rounding down (i.e., more negative)



In [67]:
import numpy

example_demo = [-4.5, -3.5, -0.05, 0.05, 3.5, 4.5]

rounding_methods = [numpy.around, numpy.ceil, numpy.floor]

for method in rounding_methods:
    for number in example_demo:
        number_rounded = method(number)

        print(f"'numpy.{method.__name__}' round number {number} to: {number_rounded}")
    print()

'numpy.around' round number -4.5 to: -4.0
'numpy.around' round number -3.5 to: -4.0
'numpy.around' round number -0.05 to: -0.0
'numpy.around' round number 0.05 to: 0.0
'numpy.around' round number 3.5 to: 4.0
'numpy.around' round number 4.5 to: 4.0

'numpy.ceil' round number -4.5 to: -4.0
'numpy.ceil' round number -3.5 to: -3.0
'numpy.ceil' round number -0.05 to: -0.0
'numpy.ceil' round number 0.05 to: 1.0
'numpy.ceil' round number 3.5 to: 4.0
'numpy.ceil' round number 4.5 to: 5.0

'numpy.floor' round number -4.5 to: -5.0
'numpy.floor' round number -3.5 to: -4.0
'numpy.floor' round number -0.05 to: -1.0
'numpy.floor' round number 0.05 to: 0.0
'numpy.floor' round number 3.5 to: 3.0
'numpy.floor' round number 4.5 to: 4.0



<div class="alert alert-block alert-warning">
<hr style="border:1.5px dashed gray"></hr>

## Take-home message

1. <font color='dodgerblue'>**Think critically**</font> about what you are doing and what your final goal is. (In other words, don't take information that you find off the Internet and blindly use it.)


2. **Create short test codes** to **verify** that <font color='dodgerblue'>**what you think you are doing**</font> is actually <font color='dodgerblue'>**what you are doing**</font>!


3. Consider using `decimals` (better rounding options and correction of float-point representation error) or `numpy` (once we get to the Numpy lectures).


4. Rounding via string formatting can cause unexpected results.